### ISSUE TITLE
Superset Login Error through OpenShift OAuth (SSLCertVerificationError)

### ISSUE THREAD

**Issue by muhyid** [opendatahub-io/odh-manifests#619]
Dear All
I just installed Open Data Hub v1.3.0 through OperatorHub into the OpenShift Container Platform v4.10.24. I configured KfDef and its kustomizeConfig parameters. All required resources (pods, config map, deployments, routes, etc.) are created properly. I opened the superset route and try to log in via OpehShift integration (Sign in with OpenShift), but I encountered the below error.
The request to sign in was denied.
When I check the logs from the superset pod I noticed those

2022-08-04 17:59:54,249:INFO:superset.utils.screenshots:No PIL installation found
2022-08-04 18:04:58,282:WARNING:root:Class 'werkzeug.local.LocalProxy' is not mapped
**2022-08-04 18:05:24,465:ERROR:flask_appbuilder.security.views:Error authorizing OAuth access token: HTTPSConnectionPool(host='oauth-openshift.apps.openshift01.mycompany.com', port=443): Max retries exceeded with url: /oauth/token (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1125)')))**

I am using global/external signer certificate (Sectigo https://sectigo.com/) on my OpenShift cluster. It seems superset can't recognize my certificate chain. I guess it is expecting the self-signed local certificate.

There was the same situation on the JupyterHub (it was installed in the same implementation via ODH) I passed this issue with a workaround (`c.OpenShiftOAuthenticator.validate_cert = False` added this parameter into the jupyterhub-cfg ConfigMap)

I tried to find a similar workaround or permanent solution for superset too. I looked this GitHub and also Superset official documentation, but I couldn't find useful information. In the meantime, I tried the below parameters one by one in superset-config-py ConfigMap

```
import ssl
ssl.SSLContext.verify_mode =ssl.VerifyMode.CERT_NONE
```
```
import ssl
ssl._create_default_https_context = ssl._create_unverified_context
```

but they didn't work. 

1. If possible, please provide me a workaround because I have been stuck in this step. I need to skip certificate validation (like JupyterHub)
2. Please prepare an external custom certificate (e.g. Sectigo, Lets Encrypt, Entrust, GlobalSign, DigiCert, etc.) chain recognition option in configuration parameters (like secret or configmap, etc.)

**Steps To Reproduce**
1. Intall ODH with Superset on OCP 4.10
2. Click on Superset route
3. Clik to Sign in with OpenSHift buton
4. See superset pod logs

**Expected behavior**
You will see this error:
2022-08-04 18:05:24,465:ERROR:flask_appbuilder.security.views:Error authorizing OAuth access token: HTTPSConnectionPool(host='oauth-openshift.apps.openshift01.mycompany.com', port=443): Max retries exceeded with url: /oauth/token (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] **certificate verify failed: unable to get local issuer certificate** (_ssl.c:1125)')))

**Workaround (if any)**
I need to skip certificate validation (like JupyterHub)

**Open Data Hub Version**
ODH 1.3.0 from OperatorHub
Apache Superset v1.4.1 from ODH

**OpenShift Version**
Version #: 4.10.24 
Provider (Baremetal, OpenStack, RHV, AWS, OKD, CodeReady Containers, ...): Vmware vSphere

**Comment by muhyid** [comment]
Dear All,
Any comment about that?

**Comment by accorvin** [comment]
@muhyid the superset auth code I added is definitely in need of better debugging features. Can you try manually adding some more debugging logic in your deployment? I think you could change the `auth_user_oauth` function (defined [here](https://github.com/opendatahub-io/odh-manifests/blob/master/superset/base/superset_config.py#L139)) in the `superset-config` configmap so that it looks like the following:

```
def auth_user_oauth(self, userinfo):
    try:
        user = super(CustomSecurityManager, self).auth_user_oauth(userinfo)
        for rk in userinfo['role_keys']:
                role = self.find_role(rk)
                if role is not None and role not in user.roles:
                user.roles.append(role)
        self.update_user(user)  # update user roles
    except:
        import traceback
        import logging
        logger = logging.getLogger(self.__class__.__name__)A
        logging.error(traceback.format_exception())

    return user
```

I think that should give us better information in the superset pod log when the auth fails to understand exactly where it's failing.

**Comment by muhyid** [comment]
Hi @accorvin 

Sorry for the huge delay in replying to your suggestion. I have a lot of urgent :( tasks. 

I added your shared code block to the **superset-config-py** configmap.

**Default Value:**
![Default Config](https://user-images.githubusercontent.com/83450450/188639334-a0d46a1f-86f4-4e80-a0f8-6b06490a9131.JPG)

**Your Suggestion:**
![image](https://user-images.githubusercontent.com/83450450/188640577-ced8c0c7-ca55-4e00-8762-a6faf7f7638a.png)

And tried to recrate the pod with new config. It throwed below error

> Failed to import config for SUPERSET_CONFIG_PATH=/etc/superset/superset_config.py
Traceback (most recent call last):
  File "/opt/app-root/lib64/python3.8/site-packages/superset/config.py", line 1288, in <module>
    override_conf = imp.load_source("superset_config", cfg_path)
  File "/usr/lib64/python3.8/imp.py", line 171, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 702, in _load
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 779, in exec_module
  File "<frozen importlib._bootstrap_external>", line 916, in get_code
  File "<frozen importlib._bootstrap_external>", line 846, in source_to_code
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/etc/superset/superset_config.py", line 150
    logger = logging.getLogger(self.__class__.__name__)A
                                                       ^
SyntaxError: invalid syntax
Failed to create app

When I delete the added lines, the pod created successfully

After that, I added your suggested code block again but deleted the "logger" line last character A `(logger = logging.getLogger(self.__class__.__name__)).` The pod created successfully. I tried to login again, and I got the same error without extra information

> [2022-09-06 13:19:44 +0000] [1] [INFO] Starting gunicorn 20.1.0
[2022-09-06 13:19:44 +0000] [1] [INFO] Listening at: http://0.0.0.0:8088 (1)
[2022-09-06 13:19:44 +0000] [1] [INFO] Using worker: sync
[2022-09-06 13:19:44 +0000] [12] [INFO] Booting worker with pid: 12
[2022-09-06 13:19:44 +0000] [13] [INFO] Booting worker with pid: 13
/opt/app-root/lib64/python3.8/site-packages/urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 'openshift.default.svc.cluster.local'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
warnings.warn(
/opt/app-root/lib64/python3.8/site-packages/urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 'openshift.default.svc.cluster.local'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
warnings.warn(
Loaded your LOCAL configuration at [/etc/superset/superset_config.py]
logging was configured successfully
2022-09-06 13:19:45,905:INFO:superset.utils.logging_configurator:logging was configured successfully
Loaded your LOCAL configuration at [/etc/superset/superset_config.py]
logging was configured successfully
2022-09-06 13:19:45,911:INFO:superset.utils.logging_configurator:logging was configured successfully
2022-09-06 13:19:45,912:INFO:root:Configured event logger of type <class 'superset.utils.log.DBEventLogger'>
/opt/app-root/lib64/python3.8/site-packages/flask_caching/__init__.py:201: UserWarning: Flask-Caching: CACHE_TYPE is set to null, caching is effectively disabled.
warnings.warn(
2022-09-06 13:19:45,916:INFO:root:Configured event logger of type <class 'superset.utils.log.DBEventLogger'>
/opt/app-root/lib64/python3.8/site-packages/flask_caching/__init__.py:201: UserWarning: Flask-Caching: CACHE_TYPE is set to null, caching is effectively disabled.
warnings.warn(
No PIL installation found
2022-09-06 13:19:46,136:INFO:superset.utils.screenshots:No PIL installation found
No PIL installation found
2022-09-06 13:19:46,145:INFO:superset.utils.screenshots:No PIL installation found
2022-09-06 13:24:31,758:WARNING:root:Class 'werkzeug.local.LocalProxy' is not mapped

**2022-09-06 13:25:28,134:ERROR:flask_appbuilder.security.views:Error authorizing OAuth access token: HTTPSConnectionPool(host='oauth-openshift.apps.openshift01.mycompany.com', port=443): Max retries exceeded with url: /oauth/token (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1125)')))**


![image](https://user-images.githubusercontent.com/83450450/188648038-b8677ae7-f516-46b2-b873-6726dbc5b088.png)


Kind Regards

**Comment by accorvin** [comment]
@muhyid sorry for the typo "A" at the end of the line. Glad you caught it.

Can you try instead adding the same try/except block (i.e. adding error logging in the except) around the full contents of the oauth_user_info fuction (defined here: https://github.com/opendatahub-io/odh-manifests/blob/master/superset/base/superset_config.py#L113 )?

**Comment by muhyid** [comment]
Hi @accorvin 

**This is default setting:**

![image](https://user-images.githubusercontent.com/83450450/188838299-75ba266d-caf1-488f-97e5-2067d60541a8.png)

**This is your suggestion, Is there any wrong thing?**

![image](https://user-images.githubusercontent.com/83450450/188841464-c7763c68-33f6-4417-8e5c-9bcdfba02767.png)

I tried your suggested configuration 

> [2022-09-07 09:42:21 +0000] [1] [INFO] Starting gunicorn 20.1.0
[2022-09-07 09:42:21 +0000] [1] [INFO] Listening at: http://0.0.0.0:8088 (1)
[2022-09-07 09:42:21 +0000] [1] [INFO] Using worker: sync
[2022-09-07 09:42:21 +0000] [12] [INFO] Booting worker with pid: 12
[2022-09-07 09:42:21 +0000] [13] [INFO] Booting worker with pid: 13
/opt/app-root/lib64/python3.8/site-packages/urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 'openshift.default.svc.cluster.local'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
warnings.warn(
Loaded your LOCAL configuration at [/etc/superset/superset_config.py]
logging was configured successfully
2022-09-07 09:42:22,600:INFO:superset.utils.logging_configurator:logging was configured successfully
2022-09-07 09:42:22,606:INFO:root:Configured event logger of type <class 'superset.utils.log.DBEventLogger'>
/opt/app-root/lib64/python3.8/site-packages/flask_caching/__init__.py:201: UserWarning: Flask-Caching: CACHE_TYPE is set to null, caching is effectively disabled.
warnings.warn(
/opt/app-root/lib64/python3.8/site-packages/urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 'openshift.default.svc.cluster.local'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
warnings.warn(
Loaded your LOCAL configuration at [/etc/superset/superset_config.py]
logging was configured successfully
2022-09-07 09:42:22,634:INFO:superset.utils.logging_configurator:logging was configured successfully
2022-09-07 09:42:22,643:INFO:root:Configured event logger of type <class 'superset.utils.log.DBEventLogger'>
/opt/app-root/lib64/python3.8/site-packages/flask_caching/__init__.py:201: UserWarning: Flask-Caching: CACHE_TYPE is set to null, caching is effectively disabled.
warnings.warn(
No PIL installation found
2022-09-07 09:42:22,850:INFO:superset.utils.screenshots:No PIL installation found
No PIL installation found
2022-09-07 09:42:22,900:INFO:superset.utils.screenshots:No PIL installation found
2022-09-07 09:43:31,510:WARNING:root:Class 'werkzeug.local.LocalProxy' is not mapped

**2022-09-07 09:44:06,225:ERROR:flask_appbuilder.security.views:Error authorizing OAuth access token: HTTPSConnectionPool(host='oauth-openshift.apps.openshift01.mycompany.com', port=443): Max retries exceeded with url: /oauth/token (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1125)')))**

It seems there is no extra information.

![image](https://user-images.githubusercontent.com/83450450/188848181-cdb547a2-4524-4818-8db2-0f3ad3d59c09.png)

**Comment by accorvin** [comment]
@muhyid this is a puzzler.

We have this block in the Superset config: https://github.com/opendatahub-io/odh-manifests/blob/master/superset/base/superset_config.py#L50

This creates a new cert bundle file in the Superset pod at `/tmp/superset-combined-cert-bundle.crt` that should contain all of the certs that the openshift cluster passes by default to the pod. It _should_ contain your cluster cert chain. Could you take a look in it to see whether it does?

If it doesn't, we could try modifying the code to get your cert into there and see oauth succeeds then. If it does, we can figure out a way to get it there in a clean way.

**Comment by muhyid** [comment]
Hi @accorvin 
I looked into the `/tmp/superset-combined-cert-bundle.crt` file with grep. 
I searched the beginning line words and end-of-line words of my certificate. 
There is a strange thing, it seems my certificate (private certificate e.g. access.redhat.com) is in the bundle.crt file, but issuer certificates don't exist. 
I mean there is no Root CA and Intermediate CA (Sectigo and its intermediate) in the bundle
I guess the certificate chain cannot be completed and the signer cannot be recognized, therefore, we are getting `certificate verify failed: **unable to get local issuer certificate** (_ssl.c:1125)')))` error. 

They don't exist in the bundle.crt

![image](https://user-images.githubusercontent.com/83450450/189328910-d7ef33ad-23e7-4c5e-a9cd-f376a021986c.png)

**Comment by accorvin** [comment]
@muhyid I am about to be out on leave until November 14th so won't be able to help drive this forward until I get back. Here is what I would recommend:

* Create a kustomize overlay for your Superset deployment to add your cert chain in a configmap/secret mount on the superset pod. You can find an example of extending the superset pod here: https://github.com/AICoE/internal-data-hub/blob/main/kfdefs/base/superset/kustomization.yaml#L10
* Extend the superset configuration to add additional custom python. There are instructions at the bottom of the readme here: https://github.com/opendatahub-io/odh-manifests/blob/master/superset/README.md (search for SUPERSET_ADDITIONAL_CONFIG)
* In your custom python, you'll need to combine your mounted certificate file with the contents of `/tmp/superset-combined-cert-bundle.crt` and write the result to a new file in tmp. Finally, you'll need to set the value of the `CURL_CA_BUNDLE` environment variable to that new file (you can see where we're setting that value today at https://github.com/opendatahub-io/odh-manifests/blob/master/superset/base/superset_config.py#L57

I'm hoping that if you do that then the oauth code will use the correct cert for talking to the cluster. Let me know.

**Comment by muhyid** [comment]
Hi @accorvin 
First of all, I would like to say thank you very much for your excellent support. Have e fun on your holiday. I made a workaround for this and It worked. I added Sectigo Root and Intermediate certificates in the` /tmp/superset-combined-cert-bundle.crt` file. I will use this workaround for a while, when I have more time I will try your suggestion.

![image](https://user-images.githubusercontent.com/83450450/189483270-1aac5131-001a-4633-a4b4-469896545093.png)
